client/src/main/scala/skuber/api/WatchSource.scala

```diff
@@ -0,0 +1,121 @@
+      )
+
+      val httpFlow: Flow[(Request[O], HttpRequest), StreamElement[O], _] =
+        Flow[(Request[O], HttpRequest)].mapAsync(1) { request =>
+          rc.invoke(request._2, watch = true).map(response => (request._1, response))
+        }.mapMaterializedValue(_ => NotUsed).flatMapConcat {
+          case (r, HttpResponse(StatusCodes.OK, _, entity, _)) =>
+            val start: Source[StreamElement[O], _] = Source.single(r)
+            val results: Source[StreamElement[O], _] = bytesSourceToWatchEventSource(entity.dataBytes, bufSize)
+            val end: Source[StreamElement[O], _] = Source.single(End[O]())
+            start ++ results ++ end
+          case (r, HttpResponse(sc, _, entity, _)) =>
+            throw new K8SException(Status(message = Some("Error watching resource"), code = Some(sc.intValue())))
+        }
+
+      val outboundFlow: Flow[StreamElement[O], WatchEvent[O], NotUsed] =
+        Flow[StreamElement[O]]
+          .filter(_.isInstanceOf[Result[O]])
+          .map{
+            case Result(_, event) => event
+            case _ => throw new K8SException(Status(message = Some("Error processing watch events.")))
+          }
+
+      val feedbackFlow: Flow[StreamElement[O], (Request[O], HttpRequest), NotUsed] =
+        Flow[StreamElement[O]].scan(StreamContext(None, Waiting)){(cxt, next) =>
+          next match {
+            case Request(rv) => StreamContext(rv, Processing)
+            case Result(rv, _) => StreamContext(Some(rv), Processing)
+            case End() => cxt.copy(state = Finished)
+          }
+        }.filter(_.state == Finished).map { acc =>
+          (Request[O](acc.currentResourceVersion), createRequest[O](rc, name, acc.currentResourceVersion, rd, maxWatchRequestTimeout))
+        }
+
+      val init = b.add(initSource)
+      val http = b.add(httpFlow)
+      val merge = b.add(Merge[(Request[O], HttpRequest)](2))
+      val broadcast = b.add(Broadcast[StreamElement[O]](2))
+      val outbound = b.add(outboundFlow)
+      val feedback = b.add(feedbackFlow)
+
+      // format: OFF
+      init ~> merge ~> http     ~> broadcast ~> outbound
+              merge <~ feedback <~ broadcast
+      // format: ON
+
+      SourceShape(outbound.out)
+    })
+  }
+
+  private def createRequest[O <: ObjectResource](rc: RequestContext, name: Option[String], sinceResourceVersion: Option[String], rd: ResourceDefinition[O], timeout: Duration) = {
+    rc.buildRequest(HttpMethods.GET, rd, name, query = createQueryParams(sinceResourceVersion, timeout), watch = true)
+  }
+
+  private def createQueryParams[O <: ObjectResource](sinceResourceVersion: Option[String], timeout: Duration) = {
+    Some(Uri.Query("timeoutSeconds" -> timeout.toSeconds.toString :: sinceResourceVersion.map(v => "resourceVersion" -> v).toList: _*))
+  }
+
+  private def bytesSourceToWatchEventSource[O <: ObjectResource](bytesSource: Source[ByteString, _],
```

Comment:
This largely overlaps with https://github.com/doriordan/skuber/blob/master/client/src/main/scala/skuber/api/Watch.scala#L91, could we refactor a little to reuse the same basic json parsing from both watchers?